$OpenBSD: patch-libsoup_soup-server_c,v 1.1.2.1 2011/07/28 18:10:50 jasper Exp $

* CVE-2011-2054: Fixed a security hole that caused some
  SoupServer users to unintentionally allow accessing the
  entire local filesystem when they thought they were only
  providing access to a single directory. [#653258] 

--- libsoup/soup-server.c.orig	Tue Nov 23 16:45:20 2010
+++ libsoup/soup-server.c	Thu Jul 28 20:08:22 2011
@@ -770,6 +770,15 @@ got_headers (SoupMessage *req, SoupClientContext *clie
 
 		uri = soup_message_get_uri (req);
 		decoded_path = soup_uri_decode (uri->path);
+
+		if (strstr (decoded_path, "/../") ||
+		    g_str_has_suffix (decoded_path, "/..")) {
+			/* Introducing new ".." segments is not allowed */
+			g_free (decoded_path);
+			soup_message_set_status (req, SOUP_STATUS_BAD_REQUEST);
+			return;
+		}
+
 		soup_uri_set_path (uri, decoded_path);
 		g_free (decoded_path);
 	}
